File: client-sdk/go/maxine_client.go
package maxine

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net"
	"net/http"
	"time"
)

type MaxineClient struct {
	BaseURL string
}

type ServiceNode struct {
	NodeName string `json:"nodeName"`
	Address  string `json:"address"`
}

type RegisterPayload struct {
	ServiceName string                 `json:"serviceName"`
	NodeName    string                 `json:"nodeName"`
	HostName    string                 `json:"hostName"`
	Port        int                    `json:"port"`
	Weight      int                    `json:"weight,omitempty"`
	Metadata    map[string]interface{} `json:"metadata,omitempty"`
}

func NewMaxineClient(baseURL string) *MaxineClient {
	return &MaxineClient{BaseURL: baseURL}
}

func (c *MaxineClient) Register(payload RegisterPayload) error {
	url := c.BaseURL + "/api/maxine/serviceops/register"
	data, err := json.Marshal(payload)
	if err != nil {
		return err
	}
	resp, err := http.Post(url, "application/json", bytes.NewBuffer(data))
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		body, _ := ioutil.ReadAll(resp.Body)
		return fmt.Errorf("registration failed: %s", string(body))
	}
	return nil
}

func (c *MaxineClient) Discover(serviceName string) (*ServiceNode, error) {
	url := c.BaseURL + "/api/maxine/serviceops/discover?serviceName=" + serviceName + "&proxy=false"
	resp, err := http.Get(url)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		body, _ := ioutil.ReadAll(resp.Body)
		return nil, fmt.Errorf("discovery failed: %s", string(body))
	}
	var node ServiceNode
	err = json.NewDecoder(resp.Body).Decode(&node)
	return &node, err
}

func (c *MaxineClient) DiscoverUDP(serviceName string, udpPort int, udpHost string) (*ServiceNode, error) {
	conn, err := net.Dial("udp", fmt.Sprintf("%s:%d", udpHost, udpPort))
	if err != nil {
		return nil, err
	}
	defer conn.Close()
	_, err = conn.Write([]byte(serviceName))
	if err != nil {
		return nil, err
	}
	buffer := make([]byte, 1024)
	n, err := conn.Read(buffer)
	if err != nil {
		return nil, err
	}
	var node ServiceNode
	err = json.Unmarshal(buffer[:n], &node)
	return &node, err
}

func (c *MaxineClient) DiscoverTCP(serviceName string, tcpPort int, tcpHost string) (*ServiceNode, error) {
	conn, err := net.Dial("tcp", fmt.Sprintf("%s:%d", tcpHost, tcpPort))
	if err != nil {
		return nil, err
	}
	defer conn.Close()
	_, err = conn.Write([]byte(serviceName + "\n"))
	if err != nil {
		return nil, err
	}
	buffer := make([]byte, 1024)
	n, err := conn.Read(buffer)
	if err != nil {
		return nil, err
	}
	var node ServiceNode
	err = json.Unmarshal(bytes.TrimSpace(buffer[:n]), &node)
	return &node, err
}

func (c *MaxineClient) Deregister(serviceName, nodeName string) error {
	url := c.BaseURL + "/api/maxine/serviceops/deregister"
	payload := map[string]string{"serviceName": serviceName, "nodeName": nodeName}
	data, err := json.Marshal(payload)
	if err != nil {
		return err
	}
	req, err := http.NewRequest("DELETE", url, bytes.NewBuffer(data))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		body, _ := ioutil.ReadAll(resp.Body)
		return fmt.Errorf("deregistration failed: %s", string(body))
	}
	return nil
}

func (c *MaxineClient) SendHeartbeat(payload RegisterPayload) {
	go func() {
		for {
			err := c.Register(payload)
			if err != nil {
				fmt.Printf("Heartbeat failed: %v\n", err)
			}
			time.Sleep(5 * time.Second)
		}
	}()
}

File: client-sdk/csharp/MaxineClient.cs
using System;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;

namespace Maxine
{
    /// <summary>
    /// C# client for Maxine service registry
    /// </summary>
    public class MaxineClient : IDisposable
    {
        private readonly HttpClient _httpClient;
        private readonly string _registryUrl;

        public MaxineClient(string registryUrl)
        {
            _registryUrl = registryUrl.TrimEnd('/');
            _httpClient = new HttpClient();
        }

        /// <summary>
        /// Discover a service node
        /// </summary>
        /// <param name="serviceName">The service name to discover</param>
        /// <returns>ServiceNode or null if not found</returns>
        public async Task<ServiceNode?> DiscoverAsync(string serviceName)
        {
            return await DiscoverAsync(serviceName, null, null, null);
        }

        /// <summary>
        /// Discover a service node with parameters
        /// </summary>
        /// <param name="serviceName">The service name</param>
        /// <param name="namespace">The namespace (default: default)</param>
        /// <param name="version">The version</param>
        /// <param name="proxy">Whether to proxy or return address</param>
        /// <returns>ServiceNode or null if not found</returns>
        public async Task<ServiceNode?> DiscoverAsync(string serviceName, string? @namespace, string? version, bool? proxy)
        {
            try
            {
                string url = $"{_registryUrl}/api/maxine/serviceops/discover?serviceName={serviceName}";
                if (@namespace != null) url += $"&namespace={@namespace}";
                if (version != null) url += $"&version={version}";
                if (proxy != null) url += $"&proxy={proxy}";

                HttpResponseMessage response = await _httpClient.GetAsync(url);
                if (response.IsSuccessStatusCode)
                {
                    string json = await response.Content.ReadAsStringAsync();
                    var node = JsonSerializer.Deserialize<ServiceNode>(json);
                    return node;
                }
                else
                {
                    Console.WriteLine($"Service discovery failed for {serviceName}: {response.StatusCode}");
                    return null;
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error discovering service {serviceName}: {ex.Message}");
                return null;
            }
        }

        /// <summary>
        /// Discover a service node via UDP for ultra-fast lookups
        /// </summary>
        /// <param name="serviceName">The service name to discover</param>
        /// <param name="udpPort">The UDP port of the Maxine server</param>
        /// <param name="udpHost">The UDP host of the Maxine server</param>
        /// <returns>ServiceNode or null if not found</returns>
        public async Task<ServiceNode?> DiscoverUDPAsync(string serviceName, int udpPort = 8081, string udpHost = "localhost")
        {
            try
            {
                using var udpClient = new System.Net.Sockets.UdpClient();
                udpClient.Client.ReceiveTimeout = 1000; // 1 second timeout
                var sendBytes = System.Text.Encoding.UTF8.GetBytes(serviceName);
                await udpClient.SendAsync(sendBytes, sendBytes.Length, udpHost, udpPort);

                var receiveResult = await udpClient.ReceiveAsync();
                var response = System.Text.Encoding.UTF8.GetString(receiveResult.Buffer);
                var node = System.Text.Json.JsonSerializer.Deserialize<ServiceNode>(response);
                return node;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error discovering service {serviceName} via UDP: {ex.Message}");
                return null;
            }
        }

        /// <summary>
        /// Discover a service node via TCP for ultra-fast lookups
        /// </summary>
        /// <param name="serviceName">The service name to discover</param>
        /// <param name="tcpPort">The TCP port of the Maxine server</param>
        /// <param name="tcpHost">The TCP host of the Maxine server</param>
        /// <returns>ServiceNode or null if not found</returns>
        public async Task<ServiceNode?> DiscoverTCPAsync(string serviceName, int tcpPort = 8082, string tcpHost = "localhost")
        {
            try
            {
                using var tcpClient = new System.Net.Sockets.TcpClient();
                await tcpClient.ConnectAsync(tcpHost, tcpPort);
                tcpClient.ReceiveTimeout = 1000; // 1 second timeout

                var stream = tcpClient.GetStream();
                var sendBytes = System.Text.Encoding.UTF8.GetBytes(serviceName + "\n");
                await stream.WriteAsync(sendBytes, 0, sendBytes.Length);

                var buffer = new byte[1024];
                var bytesRead = await stream.ReadAsync(buffer, 0, buffer.Length);
                var response = System.Text.Encoding.UTF8.GetString(buffer, 0, bytesRead).Trim();
                var node = System.Text.Json.JsonSerializer.Deserialize<ServiceNode>(response);
                return node;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error discovering service {serviceName} via TCP: {ex.Message}");
                return null;
            }
        }

        public void Dispose()
        {
            _httpClient.Dispose();
        }
    }

    /// <summary>
    /// Service node representation
    /// </summary>
    public class ServiceNode
    {
        public string Address { get; set; } = "";
        public string NodeName { get; set; } = "";

        public override string ToString()
        {
            return $"ServiceNode{{Address='{Address}', NodeName='{NodeName}'}}";
        }
    }
}